### PR TITLE
Tag energy boundaries as pymoab sparse

### PR DESCRIPTION
**Pull request by zxkjack123**
##  Description
Add energy boundaries data (`e_bins`) as a PyMOAB SPARSE tag. This PR is the replacement of PR #1157 .

## Motivation and Context
This PR aims to tag energy boundaries data (`e_bins`) as a PyMOAB NativeMeshTag for a mesh in the r2s workflow. The `e_bins` tag will replace the file "e_bounds" in the photon transport step of PyNE R2S. 

## Behaviour
Before this PR, a "e_bounds" text file is required in photon transport of r2s.
After this PR, the "e_bounds" is no longer required, the `e_bounds` information can be passed in by `source.h5m` file. The `e_bounds` are stored in the `source.h5m` as a sparse tag.
For compability, it is still supported to use text file with the old `source.h5m` generated before this PR.

## Todo
- [x] : read `e_bounds` from h5m file during the source sampling
- [x] : add new sampler without e_bounds argument
- [x] : compatible of two ways to pass in `e_bounds`:
    - [x] : use 'e_bounds' in `source.h5m` if it exists
    - [x] : use 'e_bounds' provided by txt file if 'source.h5m' does not contain it 
- [x] : update docs

### CODE REVIEW

**Comment by zxkjack123** [comment]
@pshriwise I am trying to tag a sparse tag on mesh with `NativeMeshTag` with `MB_TAG_SPARSE`. The data is tagged, but it works like `MB_TAG_DNESE`. Do I use the wrong parameters here?

**Comment by pshriwise** [comment]
This means it behaves like a dense tag in that each element has a tag value after the tag is created?

**Comment by pshriwise** [comment]
When you say it works like `MB_TAG_DENSE` this means that all of the elements have a value when you're only trying to tag a few?

**Comment by zxkjack123** [comment]
> When you say it works like `MB_TAG_DENSE` this means that all of the elements have a value when you're only trying to tag a few?

Yes. I want the tag data appears only once, but it appears `num_ves` times, just like `MB_TAG_DENSE` tag. For example, the data I want to tag in the test function is `[0.0, 0.1, 20]`, the targe mesh has 4 elements. However, the output tag data is: 
```
[[0.0, 0.1, 20],
[0.0, 0.1, 20],
[0.0, 0.1, 20],
[0.0, 0.1, 20]]
```

**Comment by pshriwise** [comment]
Sounds good. I'll take a look shortly.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py:997]: I think the storage types listed below should be `'sparse'` and `'dense'` based on the type description here and the code below.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]: I think we only want to check if the `storage_type` is equal to `'sparse'` here.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]: This is a nice convenience function, but it seems a little out of place, a little too specific, for this generic `Mesh` class. It might be better off in the location of the code being used?

File: tests/test_mesh.py
Comment (pshriwise) [inline tests/test_mesh.py]: This test looks good to me. Based on my comment above we may want to create the tag manually here instead.

**Review by pshriwise — CHANGES_REQUESTED** [review]
Thanks for the addition @zxkjack123! You did a nice job keeping the internals of the Mesh class' reliance on PyMOAB mostly separate from the external API. I just have a few comments on how to complete that and on the `tag_e_ bins` convenience method.

File: tests/test_mesh.py
Comment (zxkjack123) [reply tests/test_mesh.py]: @pshriwise I have moved the `tag_e_bins` to `r2s` part as it is used in `r2s step2`. The test function fails for me (with pymoab version '5.1.0'), here is the error message:
```
FAIL: test_r2s.test_tag_e_bins
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gw/.local/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/gw/opt/pyne/tests/test_r2s.py", line 378, in test_tag_e_bins
    assert_array_equal(m.e_bins[:], e_bins)
  File "/home/gw/.local/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 931, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/home/gw/.local/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 759, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal

(shapes (4, 3), (3,) mismatch)
 x: array([[ 0. ,  0.1, 20. ],
       [ 0. ,  0.1, 20. ],
       [ 0. ,  0.1, 20. ],
       [ 0. ,  0.1, 20. ]])
 y: array([ 0. ,  0.1, 20. ])
```
How should we do it if we want to create the tag `manually`?

Comment (pshriwise) [reply tests/test_mesh.py]: Sorry for taking to long to get back to you @zxkjack123. It looks like there are a couple of things preventing the behavior you want. First, there's this behavior at the bottom of `Mesh.tag`:

https://github.com/pyne/pyne/blob/dc5fccf9b1624fc6e70e976f6c433e5a2e7f074b/pyne/mesh.py#L1020-L1021

If a tag is added to the mesh with the `value` keyword set. The method will set that value for all elements on the mesh. This is happening in the `t[:] = value` line. So even though it is being stored in a sparse manner in PyMOAB, all elements are getting data and it behaves like a dense tag.

Second, there's a similar line in the `NativeMeshTag` constructor:

https://github.com/pyne/pyne/blob/dc5fccf9b1624fc6e70e976f6c433e5a2e7f074b/pyne/mesh.py#L398-L399

which will set a default value if provided. Changing the default value to `None` should fix this problem, though if a tag is dense it's considered best practice to provide a default value. I'll leave it up to you as to how you'd like to address that.

With that fix to the second problem in place, the best way I can see to set a single tag value for the mesh in `tag_e_bins` would be

```python
def tag(m, e_bins, tag_name='e_bins'):
    m.tag(name=tag_name, doc='energy boundaries of the photon source',
              tagtype=NativeMeshTag, size=len(e_bins), dtype=float,          
              storage_type='sparse')                                                                                                                                                                                                                    
    m.get_tag(tag_name)[m] = e_bxins
    return m
```

This syntax tags the energy bins on the mesh instance.

The `test_tag_e_bins` function will need to be updated as well to only check for the energy bins on the mesh instance:

```python
assert_array_equal(m.e_bins[m], e_bins)
```

I could see that becoming a one-liner function in `r2s.py` for convenience:

```python
def get_e_bins(m):
    return m.e_bins[m]
```

**Comment by pshriwise** [comment]
Not sure what’s going on with the `changelog_update` failure. Looks like it’s getting to this line

https://github.com/pyne/pyne/blob/dc5fccf9b1624fc6e70e976f6c433e5a2e7f074b/.circleci/changelog_test.sh#L20

and failing due the value of `default_branch` being `null`. Maybe I’m misinterpreting that though.

Any thoughts @gonuke, @bam241?

**Comment by zxkjack123** [comment]
Thanks for the help of @pshriwise , this PR is ready to review.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]:
```suggestion
            raise TypeError("Expansion of sparse tags is not implemented.")
```

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]: Along the lines of this comment. I'm wondering if we would raise a warning for a non-default value of `storage_type` if the specified tag type isn't `NativeMeshTag`?

File: src/source_sampling.cpp
Comment (pshriwise) [inline src/source_sampling.cpp]: Let's remove this commented line.

File: src/source_sampling.cpp
Comment (pshriwise) [inline src/source_sampling.cpp]: Indentation updates needed here.

File: src/source_sampling.cpp
Comment (pshriwise) [inline src/source_sampling.cpp]: These data member initializers can probably do on the same line.

File: src/source_sampling.h
Comment (pshriwise) [inline src/source_sampling.h:184]:
```suggestion

    /// Constuctor for simplified overall sampler
```

File: src/source_sampling.cpp
Comment (pshriwise) [inline src/source_sampling.cpp]:
```suggestion
    rval = mesh->tag_get_data(e_bounds_tag, &root_set, 1, e_bounds.data());
```

File: pyne/source_sampling.pyx
Comment (pshriwise) [inline pyne/source_sampling.pyx:397]: Missing parameter description for the `mode` argument here.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]:
```suggestion
            dense - MB_TAG_DENSE tag, values are stored in arrays which match
                arrays of contiguous entity handles. Dense tags are more
                efficient in both storage and memory if large numbers of
                entities are assigned the same tag. Storage for a given dense
                tag is not allocated until a tag value is set on an entity, at which point
                memory allocation for the dense tag occurs for all entities.
```

Comment (pshriwise) [reply pyne/mesh.py]: The sparse description was present twice here. I also made some updates to the description of dense tags.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]: The description updates below apply here as well.

File: docs/usersguide/source_sampling.rst
Comment (pshriwise) [inline docs/usersguide/source_sampling.rst:157]: The `"source_density"` parameter was removed in the source code below. Does it need to be removed here as well?

**Review by pshriwise — CHANGES_REQUESTED** [review]
The removal of the `e_bounds` file is a significant improvement in streamlining the R2S workflow. Thanks for all your effort @zxkjack123!

Just some line comments here.

**Comment by bam241** [comment]
The Changelog test is failing because you are following the PyNE project on CI.
So CI is running on your account and not the PyNE account.
The Changelog test needs the variable set in the PyNE account to run, but can't access them.

I would suggest you unfollow the PyNE project on you personal CircleCI account, which is not required unless you want CI to run on PR you do on your own PyNE fork.
Let me know if any of this does not make sense...

File: pyne/mesh.py
Comment (zxkjack123) [reply pyne/mesh.py]: Done in https://github.com/pyne/pyne/pull/1341/commits/e9020ad9f10e21d928064808cad076131b064b1d.

File: src/source_sampling.cpp
Comment (zxkjack123) [reply src/source_sampling.cpp]: Done in https://github.com/pyne/pyne/pull/1341/commits/0315a7f4cde0e9080c8d4f77b27e4a5f3665a7e3.

File: src/source_sampling.cpp
Comment (zxkjack123) [reply src/source_sampling.cpp]: Done in https://github.com/pyne/pyne/pull/1341/commits/296650d9683e6f43ebbb2dab15ff59edfccf2593.

File: pyne/source_sampling.pyx
Comment (zxkjack123) [reply pyne/source_sampling.pyx:397]: Parameter description updated in https://github.com/pyne/pyne/pull/1341/commits/6adbeb57e8a52ead1ffe7f28c2140750edfd98a1.

File: docs/usersguide/source_sampling.rst
Comment (zxkjack123) [reply docs/usersguide/source_sampling.rst:157]: No, the "source_density" is the (hard coded) value of the parameter "src_tag_name". It is still used in the code.

**Comment by zxkjack123** [comment]
> The Changelog test is failing because you are following the PyNE project on CI.
> So CI is running on your account and not the PyNE account.
> The Changelog test needs the variable set in the PyNE account to run, but can't access them.
> 
> I would suggest you unfollow the PyNE project on you personal CircleCI account, which is not required unless you want CI to run on PR you do on your own PyNE fork.
> Let me know if any of this does not make sense...

Thanks @bam241 . Your method works.

**Review by pshriwise — APPROVED** [review]
Looks like everything has been addressed and this seems to work nicely. Thanks for your diligence @zxkjack123!

Will merge tomorrow if there are no other comments.

**Pull request merged**
